include/pybind11/pybind11.h

```diff
@@ -1344,15 +1344,27 @@ PYBIND11_NOINLINE inline void print(tuple args, dict kwargs) {
     auto sep = kwargs.contains("sep") ? kwargs["sep"] : cast(" ");
     auto line = sep.attr("join")(strings);
 
-    auto file = kwargs.contains("file") ? kwargs["file"].cast<object>()
-                                        : module::import("sys").attr("stdout");
+    object file;
+    if (kwargs.contains("file")) {
+        file = kwargs["file"].cast<object>();
+    } else {
+        try {
+            file = module::import("sys").attr("stdout");
+        } catch (const import_error &) {
+            /* If print() is called from code that is executed as
```

Comment:
That's an interesting situation. I guess that comes up with the `ConstructorStats` output from destructors.
Wouldn't it be better to add the `try/catch` to `ConstructorStats` rather than penalize every `py::print` call?

In general, any Python code executed in a destructor can throw an exception, so this isn't really isolated to `py::print`. Even with pure C++ code it's up to the destructor to make sure that no exceptions escape -- the called functions don't need to be aware if they are running from a destructor or not.